Improve cast optimizations

tlively opened the pull request:
Simplify the optimization of ref.cast and ref.test in OptimizeInstructions by
moving the loop that examines fallthrough values one at a time out to a shared
function in properties.h. Also simplify ref.cast optimization by analyzing the
cast result in just one place.

In addition to simplifying the code, also make the cast optimizations more
powerful by analyzing the nullability and heap type of the cast value
independently, resulting in a potentially more precise analysis of the cast
behavior. Also improve optimization power by considering fallthrough values when
optimizing the SuccessOnlyIfNonNull case.

tlively commented:
Current dependencies on/for this PR:
* main
  * **PR #5876** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/5876" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/5876?utm_source=stack-comment).

kripken commented on src/ir/properties.h:
```diff
@@ -362,6 +362,83 @@ inline Expression* getFallthrough(
 }
 
+// Look at all the intermediate fallthrough expressions and return the most
+// precise type we know this value will have.
```
It would be good to comment how this handles the case of not having a most precise type. From the code it looks like nullability and heap types are treated independently, but the problem can happen among heap types too? And the code looks like it does a greedy search, so it might not find the "deepest" result,
```
    A
   / \
  B   C
 /
D
```
Here we might return C when D might be better for the caller. Though even between B and C it isn't clear which the caller would prefer, and the old code considered them both - can't that be better?

tlively replied:
If we see a value typed with both `B` (or `D`) and `C`, we know the value cannot possibly exist, so if we observe that, we can jump straight to the bottom type. The commit I just pushed now optimizes this case.

kripken commented on src/ir/properties.h:
```diff
@@ -362,6 +362,92 @@ inline Expression* getFallthrough(
+  if (!type.isRef()) {
+    // Only reference types can be improved (excepting improvements to
+    // unreachable, which we leave to refinalization).
+    // TOOD: Handle tuples if that ever becomes important.
```
```suggestion
    // TODO: Handle tuples if that ever becomes important.
```

kripken commented on src/passes/OptimizeInstructions.cpp:
```diff
@@ -1995,39 +1995,60 @@ struct OptimizeInstructions
+        break;
+      case GCTypeUtils::Success:
+      case GCTypeUtils::SuccessOnlyIfNonNull: {
+        // We know the cast will succeed, so we can try to optimize it out. Find
```
This comment matches `Success` but not `SuccessIfNonNull`.

tlively replied:
Fixed.

kripken commented on src/passes/OptimizeInstructions.cpp:
```diff
@@ -1995,39 +1995,60 @@ struct OptimizeInstructions
+        auto** refp = Properties::getMostRefinedFallthrough(
+          &curr->ref, getPassOptions(), *getModule());
+        auto* ref = *refp;
+        assert(ref->type.isRef());
```
Must this be a ref? What about
```wat
(ref.cast $foo
  (block (ref $foo)
    (unreachable)
  )
)
```
?

tlively replied:
Both `getFallthroughType` and `getMostRefinedFallthrough` return early if they encounter an unreachable, so they always produce referenced types and expressions with reference types as long as their original input had a reference type.

kripken replied:
But note that in my example the type of the block is *not* unreachable, it is `(ref $foo)`, even though the contents are unreachable. I see a check in the code for the type of the block but not for its content, so I think it will return the last item in the block in that case, which is unreachable.

tlively replied:
`getFallthroughType` and `getMostRefinedFallthrough` will return `(ref $foo)` and a pointer to the block, respectively here.

kripken replied:
I see, thanks, that's an API change compared to `getFallthrough` from before that confused me I guess.

kripken commented on src/passes/OptimizeInstructions.cpp:
```diff
@@ -1995,39 +1995,60 @@ struct OptimizeInstructions
+      Properties::getFallthroughType(curr->ref, getPassOptions(), *getModule());
+
+    // If the cast admits null but we know the value is not null, then as a
+    // first step we can tighten up the cast to stop admitting nulls.
```
Can we not just apply `cast->type = refType`? That is, the cast must be as refined as the output of `getFallthroughType`, mustn't it? So we can apply not only the nullability but the heap type.

tlively replied:
We can only do that if the cast would definitely succeed (i.e. when the value's type is a subtype of the cast type), but in that case we prefer to optimize the cast out entirely, not just improve its type. Otherwise we would be changing the semantics of the cast such that it always succeeds, even if that were not originally the case. This could also cause validation failures if we ended up changing the cast type to some unrelated type.

kripken replied:
I don't understand. Maybe I'm confused about the situation here. Is this transformation not relevant and valid somehow?
```wat
(ref.cast null struct
  (block (result anyref) ;; 'for'get' about the better typing in the children
    (ref.as_non_null
      (ref.cast null $struct.foo ..))))
=>
(ref.cast $struct.foo ;; applied the refined nullability and heap
  (block (result anyref)
    (ref.as_non_null
      (ref.cast null $struct.foo ..))))
```
That seems valid since if we get to the cast then all the constraints above us must have been applied together, so we can use the intersection of them all.

tlively replied:
Yes, that example's fine, but only because we know the cast will succeed by looking at all the fallthrough values. We already do even better by just replacing the cast entirely in this case.

Here's an example that would go wrong:

```
(ref.cast null $struct
  (block (result anyref)
    (ref.cast i31 ..)))
```

This cast will fail, so if we just replaced it with `ref.cast (ref i31)`, that would be a misoptimization. If this cast is the child of a struct.get, it would also cause a validation failure.

In contrast to arbitrarily changing the heap type here, it is valid to improve the nullability from nullable to nonnullable because that always refines the type and cannot be a misoptimization or cause a validation failure. Furthermore, performing this nullability optimization is not already handled somewhere else.

kripken replied:
Oh, true, but I think we just need the intersection of it all, including the original cast? The intersection of `(ref i31)` and `(ref null $struct)` is empty, so we could put `(ref.cast none` (empty set) in your example. (And in my example, adding the intersection of the original cast has no effect.)

tlively replied:
Yes, that would work, but it would be entirely superseded by the optimizations below. In the example you gave, the cast would be removed and the contents of the block would be teed out. In my example, the cast would be turned into a drop followed by `unreachable`.

kripken replied:
It seems conceptually simpler to me, though, to just take the intersection of all the types (in fallthroughs + the original cast), rather than treat the nullability in a special way + add the code below. Do you disagree?

tlively replied:
If we had a function in wasm-types.h for taking the meet/glb/intersection of two types, then maybe using that might be simpler, but we don't have a function like that ready to go. How would you feel if I updated the comment to point out that improving the heap type here wouldn't be useful due to the later optimizations?

kripken replied:
I was thinking that the intersection is what is sort of already computed here - for nullability, improve to non-null, and for heap type, either go to nothing (if unrelated types) or pick the more refined of the two (if related). I see that code in this PR already, I think, so I was hoping there's a good way to refactor it to use that.

But maybe I'm not seeing some practical complication here that prevents it. You're closer to the code. If that seems impractical then sgtm to continue with this PR more or less as is and leave improvements for later?

tlively replied:
Ok, I think you're right that this would be a nice refactoring in general, even if the refinement of the heap type isn't independently useful in this particular case.

tlively replied:
PTAL at the last couple of commits ('glb' and 'fewer test changes') for this refactoring.

kripken commented on src/passes/OptimizeInstructions.cpp:
```diff
@@ -1995,39 +1995,60 @@ struct OptimizeInstructions
+              replaceCurrent(builder.makeRefAs(RefAsNonNull, curr->ref));
+            } else {
+              replaceCurrent(ref);
+            }
```
Can't this skip necessary code? `ref` might be a nested child with other stuff we need around it, like this:
```wat
(ref.cast
  (block
    (..side-effect..)
    (ref)
  )
)
```

tlively replied:
This is guarded by a `ref == curr->ref` check, so we know there can be nothing else in the middle.

kripken replied:
Ah, good point, I missed that.

kripken commented on src/passes/OptimizeInstructions.cpp:
```diff
@@ -2060,114 +2086,76 @@ struct OptimizeInstructions
+        // expression. The only way this can happen is if we were able to infer
+        // that the value has bottom heap type because it was typed with
+        // multiple, incompatible heap types. In this case, the cast succeeds
+        // only if the value is null, so we can fall through to that case.
```
I don't follow this paragraph, sorry. What is "value" - the immediate value to the cast, i.e. `curr->ref`?

Can you perhaps give an example of this in code? Also, can we add an assert to verify the conclusion here?

tlively replied:
Update the comment, replacing "value" with "input" to try to clarify. I also added assertions.

tlively replied:
Now added an example as well.

kripken commented on test/lit/passes/optimize-instructions-gc.wast:
```diff
@@ -2162,6 +2356,327 @@
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $compatible-cast-fallthrough-null-check (param $eqref eqref) (result (ref i31))
+    ;; Similar to above, but now the cast only succeeds if the value is non-null.
```
The previous test also added a `ref.as_non_null`? Should that test have had a nullable cast? (I don't seem to see such a test.)

tlively replied:
In the previous test, we know the value was non-null because of the `ref.as_non_null`. This test does not have the `ref.as_non_null`, so we don't know until runtime whether the value is non-null. The output is similar, but the logic to get there is slightly different.

tlively replied:
I improved the comment here.

kripken reviewed (approved):
nice, lgtm % comments + fuzzing

tlively commented on src/passes/OptimizeInstructions.cpp:
```diff
@@ -2060,114 +2086,76 @@ struct OptimizeInstructions
+        // multiple, incompatible heap types. In this case, the cast succeeds
+        // only if the value is null, so we can fall through to that case.
+      }
+        [[fallthrough]];
```
Yes, this is a bad interaction with the special indenting rules when a switch case has braces. I don't think there's anything we can change here without making either the compiler or the linter mad.

tlively commented:
I made another small improvement to optimization of ref.test using the new getLeastUpperBound and added a test for it. I also have addressed all of the review comments now. A slightly outdated version of this PR made it through 280k fuzzer iterations before I stopped it, so I'll fuzz the latest version lightly and then land this.

kripken commented on src/passes/OptimizeInstructions.cpp:
```diff
@@ -2060,114 +2086,80 @@ struct OptimizeInstructions
-          // types for other passes.
+        }
+        // If we get here, then we know that the heap type of the cast input is
+        // more refined than the heap type of best available fallthrough
```
```suggestion
        // more refined than the heap type of the best available fallthrough
```
